Add AppLink

labkey-alan opened the pull request:
#### Rationale
This PR does two major things:

1. It adds the `<AppLink>` component, which ensures that when passed a URL that points to an AppURL for the current App it uses react-router's `<Link>`, otherwise it uses a normal `<a>`, which ensures the back button always works as intended in our apps. This should resolve [Issue 53021](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=53021).
2. It adds optional `productId` and `containerPath` to AppURL. This is backwards incompatible because we default to the  _current primary app id_, previously all AppURLs would point to current container and current product only.

#### Related Pull Requests
- https://github.com/LabKey/labkey-api-js/pull/195
- https://github.com/LabKey/labkey-ui-components/pull/1790
- https://github.com/LabKey/labkey-ui-premium/pull/748
- https://github.com/LabKey/platform/pull/6639
- https://github.com/LabKey/limsModules/pull/1393
- https://github.com/LabKey/testAutomation/pull/2458

#### Changes
- AppURL: make container and product aware
  - Add setContainerPath
  - Add setProductId
  - Add fromMenuUrl static method
- Add AppLink
  - Use AppLink any time you need to render an anchor tag, even if pointing outside of our apps
- Add useAppNavigate
  - Use anytime you would otherwise use React Router's useNavigate hook
- DefaultRenderer: Use AppLink
  - Issue 53021: Grid Column Renderers can break the back button on pages using useRouteLeave
- NavItem: Use AppLink
- Remove createProductUrl
- Remove createProductUrlFromParts
- Remove createProductUrlFromPartsWithContainer
- Remove getHref

labkey-alan committed:
Add AppLink and use in DefaultRenderer

labkey-alan committed:
NavItem: Use AppLink

labkey-alan committed:
AppLink: make props optional

labkey-alan committed:
NavItem: Use AppLink

labkey-alan committed:
MenuSectionLink: Use AppLink

labkey-nicka commented on packages/components/src/internal/url/AppLink.tsx:
```diff
@@ -0,0 +1,75 @@
+        return href.replace('#', '');
+    }
+
+    const contextPath = ActionURL.getContextPath();
```
nit: Consider replacing these four calls with:

```ts
const { action, containerPath, contextPath, controller } = ActionURL.getPathFromLocation();
const baseAppUrl = `${contextPath}${containerPath}/${controller}-${action}.view#`;
```

labkey-nicka replied:
See later comment about `baseAppUrl` construction.

labkey-alan committed:
Add products.ts, getPrimaryAppProductId, update imports

labkey-alan committed:
AppLink: remove appUrl and href in favor of to

labkey-alan committed:
EffectiveRolesList: Format file

labkey-alan committed:
AppURL: Make product and container aware

labkey-alan committed:
URLResolver: use product aware AppURL

labkey-alan committed:
Add default container to test context, fix tests

labkey-alan committed:
AppURL: Add fromMenuUrl

Use AppURL.create in menu components and models

labkey-alan committed:
Remove getPicklistUrl, use AppURL instead

labkey-alan committed:
EffectiveRolesList, GroupsList: Use AppURL and AppLink

labkey-alan committed:
getStorageSectionConfig: Use product aware AppURL

labkey-alan committed:
Remove createProductUrlFromParts

labkey-alan committed:
Update release notes

labkey-alan committed:
AssayDefinitionModel.getImportURL() - Use container-aware AppURL

labkey-alan committed:
UserDetailsPanel: Use AppLink, don't use inline handlers

labkey-alan committed:
Bump @labkey/api

labkey-alan committed:
Bump api-js

labkey-alan committed:
PicklistModals/Buttons: Stop using product id

labkey-alan committed:
AppURL: support arrays, booleans, and numbers in params
 - use ActionUrl.queryString
 - remove unused includeEmptyParams flag

labkey-alan committed:
AppLink: always render anchor tag during tests
- <Link> does not render in tests

labkey-alan committed:
Add @labkey/api mock, allow tests to mock ActionURL values

labkey-alan committed:
Remove createProductUrlFromPartsWithContainer

labkey-alan committed:
Remove createProductUrl

labkey-alan committed:
update components.md

labkey-alan committed:
ProductMenuSection: remove currentProductId

labkey-alan committed:
ProductMenuSection: Fix issue with requests section

labkey-alan committed:
Remove getHref

labkey-alan committed:
UserDetailsPanel.test.tsx - cleanup

labkey-alan committed:
getJobCreationHref: return AppURL

labkey-alan committed:
AssayResultsForSamplesButton: drop current/target product id props

labkey-alan committed:
Fix mock API action value

labkey-alan committed:
Fix tests

labkey-alan committed:
MenuSectionModel: remove currentProductId

labkey-alan committed:
AppURL: Fix issue with fromMenuUrl

labkey-alan committed:
Add TODO to dropdowns.tsx

labkey-alan committed:
Fix issue with NavigationAPIWrapper.ts

labkey-nicka commented on packages/components/src/internal/AssayDefinitionModel.ts:
```diff
@@ -122,48 +122,35 @@ export class AssayDefinitionModel extends ImmutableRecord({
-        targetProductId?: string,
-        ignoreFilter?: boolean,
-        containerPath?: string
+        filterList?: Filter.IFilter[],
```
nit: Rename local variable to `filters`.

labkey-alan committed:
parseAppPath use ActionURL.buildURL

labkey-nicka commented on packages/components/src/internal/components/entities/AssayResultsForSamplesButton.spec.tsx:
```diff
@@ -37,17 +37,6 @@ describe('AssayResultsForSamplesButton', () => {
         wrapper.unmount();
     });
 
```
Given that we're touching this component the associated `AssayResultsForSamplesButton.spec.tsx` should be updated to RTL.

labkey-alan committed:
Add getPathFromLocation to ActionURL mock

labkey-alan committed:
Remove getJobURL prop

labkey-nicka commented on packages/components/src/internal/components/domainproperties/models.test.ts:
```diff
@@ -1016,15 +1016,9 @@ describe('DomainField', () => {
 
         field = field.merge({ dataType: LOOKUP_TYPE }) as DomainField;
         expect(field.getDetailsArray().slice(0, 2).join('')).toBe('Updated. Current Folder > exp');
```
Is there a different comparison we can do here? Are these no longer links?

labkey-alan replied:
The tests are now executing in an environment that appears as SampleManager and this stuff is not rendered.

labkey-nicka commented on packages/components/src/internal/url/AppURL.ts:
```diff
@@ -171,61 +135,112 @@ export class AppURL {
-    toQuery(urlPrefix?: string): Record<string, string> {
-        const query = { ...this._params };
+    getContainerPath(): string {
+        if (this._containerPath === undefined) return ActionURL.getContainer();
```
nit: Consider:

```ts
getContainerPath(): string {
    return this._containerPath ?? ActionURL.getContainer();
}
```

labkey-nicka commented on packages/components/src/internal/url/AppURL.ts:
```diff
@@ -171,61 +135,112 @@ export class AppURL {
 
-        return query;
+    getProductId(): string {
+        if (this._productId === undefined) return getPrimaryAppProductId() ?? '';
```
nit: Consider:

```ts
getProductId(): string {
    return this._productId ?? getPrimaryAppProductId() ?? '';
}
```

labkey-nicka commented on packages/components/src/internal/url/AppURL.ts:
```diff
@@ -171,61 +135,112 @@ export class AppURL {
+        const queryString = ActionURL.queryString(allParams);
+        if (queryString) appPath += `?${queryString}`;
+
+        // TODO: do not merge this FM code, it's only useful while cleaning up FM usages
```
Marker comment for TODO.

labkey-nicka commented on packages/components/src/internal/url/AppURL.ts:
```diff
@@ -171,61 +135,112 @@ export class AppURL {
+     * Returns true if the AppURL points to a path for the current app in the current container
+     */
+    isAppPath(): boolean {
+        const currentContainerPath = ActionURL.getContainer();
```
(Tangential) It's a bit unfortunate how `ActionURL` is set up to compute these "properties" of the URL. It ends up needing to run through `getPathFromLocation()` for each call making these "getters" not cheap (compared a normal property getter).

labkey-nicka commented on packages/components/src/internal/components/permissions/GroupsList.tsx:
```diff
@@ -40,23 +38,16 @@ export const GroupsList: FC<Props> = memo(props => {
-                        'admin',
-                        'groups'
-                    );
+                    const url = AppURL.create('admin', 'groups').addParams({ expand: group.value });
```
nit: Consider adding a `ADMIN_KEY = 'admin'` to app/constants.ts to be consistent with other base routes.

labkey-nicka commented on packages/components/src/internal/components/productnavigation/ProductClickableItem.tsx:
```diff
@@ -12,6 +12,7 @@ export const ProductClickableItem: FC<ProductClickableItemProps> = memo(props =>
     const onEnter = useCallback(() => setHovered(true), [setHovered]);
     const onLeave = useCallback(() => setHovered(false), [setHovered]);
 
+    // TODO: Need to use AppLinkHere, however we'd need to add several props
```
Is this a post-PR / TBD TODO? Interesting use case since these links are rendered outside of our traditional apps.

labkey-alan replied:
Yes, I think this TODO should be addressed later. Ideally we'd drop the need for onEnter/onLeave and just use CSS `:hover` to apply the different background color.

labkey-alan replied:
So, there were several other issues with this component, so I did go ahead and add support for onMouseEnter and onMouseLeave to AppLink and made this component render AppLink. However, I do think we should do a follow-on PR in the near future to get rid of the hover events, and drop support for onMouseEnter/Leave (I have added comments to the code to suggest this).

labkey-nicka commented on packages/components/src/internal/components/search/FindAndSearchDropdown.tsx:
```diff
@@ -78,16 +76,7 @@ export const FindAndSearchDropdown: FC<Props> = memo(props => {
-                        ) as string
-                    }
-                >
+                <MenuItem onClick={onSampleFinder} href={FIND_SAMPLES_BY_FILTER_HREF.toHref()}>
```
Same goes for the `FindAndSearchDropdown` with regards to updating it's associated enzyme test to RTL.

labkey-nicka commented on packages/components/src/internal/components/user/UserDetailsPanel.tsx:
```diff
@@ -294,27 +324,17 @@ export class UserDetailsPanel extends React.PureComponent<Props, State> {
-            'admin',
-            'users'
-        );
+        const manageUrl = AppURL.create('admin', 'users')
```
nit: Consider only computing links for admins:

```tsx
if (toggleDetailsModal) {
    let footer: ReactNode;
    if (user.isAdmin) {
        // We do not currently support user management in sub folders, so we create the management URL for the project
        // container.
        const manageUrl = AppURL.create('admin', 'users')
            .addParams({ usersView: 'all', 'all.UserId~eq': userId })
            .setContainerPath(project.path);

        footer = (
            <AppLink className="pull-right btn btn-default" to={manageUrl}>
                Manage
            </AppLink>
        );
    }

    return (
        <Modal
            onCancel={toggleDetailsModal}
            className="user-detail-modal"
            footer={footer}
            title={this.renderHeader()}
        >
            {this.renderBody()}
        </Modal>
    );
}
```

labkey-nicka commented on packages/components/src/internal/url/AppLink.tsx:
```diff
@@ -0,0 +1,83 @@
+    const container = ActionURL.getContainer();
+    const controller = ActionURL.getController();
+    const action = ActionURL.getAction();
+    const baseAppUrl = `${contextPath}${container}/${controller}-${action}.view#`;
```
We will want to leverage `ActionURL` to generate the `baseAppUrl` so we are getting consistent container path placement in url.

labkey-nicka commented on packages/components/src/internal/dropdowns.tsx:
```diff
@@ -342,6 +342,7 @@ export const MenuItem = forwardRef<HTMLLIElement, MenuItemProps>((props, ref) =>
         },
         [disabled, href, onClick]
     );
+    // TODO: Use AppLink (needs support for role and rel)
```
Marker for TODO.

labkey-nicka replied:
I'd agree there is merit in just having `AppLink` extend all the properties of an anchor tag.

labkey-alan replied:
I plan to do this in a separate PR for now

labkey-alan committed:
jest.config.js: Use container relative URLS, set default container

labkey-alan committed:
AppURL.test.ts - Fix buildURL tests

labkey-alan committed:
Fix unit tests

cnathe commented on packages/components/src/internal/AssayDefinitionModel.ts:
```diff
@@ -122,48 +122,35 @@ export class AssayDefinitionModel extends ImmutableRecord({
+            params = { rowId: this.id, ...params };
+            if (selectionKey) {
+                params.selectionKey = selectionKey;
+                params.dataTab = AssayUploadTabs.Grid;
```
the use case in premium enttiies/utils.tsx getImportItemsForAssayDefinitions() had `selectionKey ? AssayUploadTabs.Grid : AssayUploadTabs.Files`. Do we need to set the `params.dataTab in the else case here to match?

labkey-alan replied:
I don't think so, as I believe the default value is AssayUploadTabs.Files, plus, if we decided to change that default we'd then need to update it here and in the component, but by not setting it we leave the decision up to the component.

cnathe commented on packages/components/src/internal/components/navigation/ProductMenuSection.tsx:
```diff
@@ -97,46 +84,30 @@ export const ProductMenuSection: FC<MenuSectionProps> = memo(props => {
-            );
-        }
+        emptyLink = (
+            <AppLink to={config.emptyAppURL} className="menu-section-link">
```
do we need to setContainerPath using `containerPath` in some way here?

labkey-alan replied:
Yes, just pushed that fix, as well as fixes for the links in the product menu.

labkey-alan committed:
URLResolver: FREEZER_ITEM_SAMPLE_MAPPER don't use toHref

labkey-alan committed:
Fix api-js ActionURL.getPathFromLocation mock

labkey-alan committed:
AssayPropertiesInput.test.tsx - Fix expected URLs

labkey-alan committed:
AdvancedSettings.test.tsx - Fix submit button class

labkey-alan committed:
ReleaseNote: format file, use moduleContext, fix tests

labkey-alan committed:
UserMenuGroup.test.tsx - Fix tests

labkey-alan committed:
MenuItemModel.getUrlString - Handle originalUrl case

labkey-alan committed:
actions.ts: move getAvailableTypes below _isAvailablePropType

Relying on hoisting is a bad idea

labkey-alan committed:
actions.test.ts - Fix tests

labkey-alan committed:
app/utils.test.ts - Fix tests

labkey-alan committed:
Update test snapshots

labkey-alan committed:
InsufficientPermissionsAlert - Add class name

Easier to detect in tests

labkey-alan committed:
ManageDropdownButton - Add className to dropdown

- Useful for testing

labkey-alan committed:
auditlog/utils.ts - Fix boxes URL

labkey-alan committed:
Use BOXES_KEY constant

labkey-alan committed:
AppLink: Add title prop

labkey-alan committed:
Export QueryParamValue

labkey-alan committed:
Remove unnecessary usages of currentProductId

labkey-alan committed:
PR Feedback

labkey-alan committed:
PR Feedback

labkey-alan committed:
ProductMenuSection: drop unused iconCls, simplify dom generation

Add styles to better align text of menu headers

labkey-alan committed:
Product Navigation: fix issue with AppURLs

AppLink: add onMouseEnter, onMouseLeave
ProductClickableItem: Rename to ProductNavigationItem, use AppLink
ProductSectionsDrawer: Fix URL generation
Remove transition styles

labkey-alan committed:
AssayResultsForSamplesButton.test.tsx - Convert from spec test file

labkey-alan committed:
FindAndSearchDropdown.test.tsx - Convert from spec file

labkey-alan committed:
ProductMenuSection.test.tsx: Fix tests

labkey-alan committed:
PR Feedback

labkey-alan committed:
AppLink: Fix issue with onMouseEnter/Leave

labkey-alan committed:
Use App.WORKFLOW_KEY when generating workflow URLs

Use AppLink instead of anchor tag

labkey-alan committed:
UserDetailsPanel - Fix issue with reactivate button

labkey-alan committed:
ProductMenu.tsx - Fix click handler

labkey-alan committed:
ServerNotificationsConfig: remove unnecessary attributes

labkey-alan committed:
ServerActivityList: Convert to FC, use AppLink

labkey-alan committed:
Add useAppNavigate hook

labkey-alan committed:
Update release notes

labkey-alan committed:
ServerNotifications: Fix rendering issue with mark all as read

labkey-alan committed:
Don't render empty sections

labkey-alan committed:
Export ITEMS_KEY constant

labkey-alan committed:
Remove FM code from AppURL

labkey-alan committed:
Prep for release
